databuilder/extractor/dashboard/tableau/tableau_dashboard_utils.py

```diff
@@ -0,0 +1,220 @@
+
+        response = requests.post(url=self._metadata_url, **params)
+        return response.json()['data']
+
+    def execute(self):
+        """
+        Must be overriden by any extractor using this class. This should parse the result and yield each entity's
+        metadata one by one. 
+        """
+        pass
+
+    def extract(self):
+        """
+        Fetch one result at a time from the generator created by self.execute(), updating using the
+        static record values if needed.
+        """
+        if not self._iterator:
+                self._iterator = self.execute()
+
+        try:
+            record = next(self._iterator)
+        except StopIteration:
+            return None
+
+        if self._static_dict:
+            record.update(self._static_dict)
+
+        return record
+
+
+class TableauDashboardAuth():
+    """
+    Attempts to authenticate agains the Tableau REST API using the provided personal access token credentials.
+    When successful, it will create a valid token that must be used on all subsequent requests.
+    https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_auth.htm
+    """
+
+    API_VERSION = const.API_VERSION
+    TABLEAU_HOST = const.TABLEAU_HOST
+    SITE_NAME = const.SITE_NAME
+    TABLEAU_ACCESS_TOKEN_NAME = const.TABLEAU_ACCESS_TOKEN_NAME
+    TABLEAU_ACCESS_TOKEN_SECRET = const.TABLEAU_ACCESS_TOKEN_SECRET
+    EXCLUDED_PROJECTS = const.EXCLUDED_PROJECTS
+    EXTERNAL_CLUSTER_NAME = const.EXTERNAL_CLUSTER_NAME
+    EXTERNAL_SCHEMA_NAME = const.EXTERNAL_SCHEMA_NAME
+    CLUSTER = const.CLUSTER
+    DATABASE = const.DATABASE
+
+    def __init__(self, conf):
+        self.site_id = None
+        self._token = None
+        self._conf = conf
+        self._site_name = self._conf.get_string(TableauDashboardAuth.SITE_NAME)
+        self._tableau_host = self._conf.get_string(TableauDashboardAuth.TABLEAU_HOST)
+        self._api_version = self._conf.get_string(TableauDashboardAuth.API_VERSION)
+        self._access_token_name = self._conf.get_string(TableauDashboardAuth.TABLEAU_ACCESS_TOKEN_NAME)
+        self._access_token_secret = self._conf.get_string(TableauDashboardAuth.TABLEAU_ACCESS_TOKEN_SECRET)
+
+    @property
+    def token(self):
+        if not self._token:
+            self._token = self._authenticate()
+        return self._token
+
+    def _authenticate(self):
+        """
+        Queries the auth/signin endpoint for the given Tableau instance using a personal access token.
+        The API version differs with your version of Tableau.
+        See https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_versions.htm
+        for details or ask your Tableau server administrator.
+        """
+        self._auth_url = "https://{tableau_host}/api/{api_version}/auth/signin".format(
+            tableau_host=self._tableau_host,
+            api_version=self._api_version
+        )
+        payload = json.dumps({
+            "credentials": {
```

Comment:
Nit: inconsistent quote/double-quoting of strings. I'd say stick with single-quotes as a default.

Reply:
I agree, I've been sticking to single quotes since that's the project standard. Thanks for catching this